posthog/queries/trends.py

```diff
@@ -0,0 +1,318 @@
+            cohort = Cohort.objects.get(pk=value.replace("cohort_", ""))
+            ret_dict["label"] = "{} - {}".format(entity.name, cohort.name)
+            ret_dict["breakdown_value"] = cohort.pk
+    return ret_dict
+
+
+class Trends:
+    def _serialize_entity(self, entity: Entity, filter: Filter, team_id: int) -> List[Dict[str, Any]]:
+        if filter.interval is None:
+            filter.interval = "day"
+
+        serialized: Dict[str, Any] = {
+            "action": entity.to_dict(),
+            "label": entity.name,
+            "count": 0,
+            "data": [],
+            "labels": [],
+            "days": [],
+        }
+        response = []
+        events = process_entity_for_events(entity=entity, team_id=team_id, order_by="-timestamp",)
+        events = events.filter(filter_events(team_id, filter, entity))
+        items = aggregate_by_interval(
+            filtered_events=events,
+            team_id=team_id,
+            entity=entity,
+            filter=filter,
+            breakdown="properties__{}".format(filter.breakdown) if filter.breakdown else None,
+        )
+        for value, item in items.items():
+            new_dict = copy.deepcopy(serialized)
+            if value != "Total":
+                new_dict.update(breakdown_label(entity, value))
+            new_dict.update(append_data(dates_filled=list(item.items()), interval=filter.interval))
+            if filter.display == TRENDS_CUMULATIVE:
+                new_dict["data"] = np.cumsum(new_dict["data"])
+            response.append(new_dict)
+
+        return response
+
+    def calculate_trends(self, filter: Filter, team_id: int, actions: QuerySet) -> List[Dict[str, Any]]:
+        entities_list = []
+
+        if len(filter.entities) == 0:
+            # If no filters, automatically grab all actions and show those instead
+            filter.entities = [
+                Entity({"id": action.id, "name": action.name, "type": TREND_FILTER_TYPE_ACTIONS,}) for action in actions
+            ]
+
+        if not filter.date_from:
+            filter._date_from = (
+                Event.objects.filter(team_id=team_id)
+                .order_by("timestamp")[0]
+                .timestamp.replace(hour=0, minute=0, second=0, microsecond=0)
+                .isoformat()
+            )
+        if not filter.date_to:
+            filter._date_to = now().isoformat()
+
+        for entity in filter.entities:
+            if entity.type == TREND_FILTER_TYPE_ACTIONS:
+                try:
+                    db_action = [action for action in actions if action.id == entity.id][0]
+                    entity.name = db_action.name
+                except IndexError:
+                    continue
+            entities_list.extend(
+                handle_compare(entity=entity, filter=filter, func=self._serialize_entity, team_id=team_id)
+            )
+
+        return entities_list
+
+    def run(self, filter: Filter, team: Team, actions: QuerySet):
```

Comment:
This pattern helps a lot to unload logic from the /api. 

I think for the case of reusability it may help to provide detailed comments above some of the utility functions (such as describing the params needed and the shape of the output). Mainly suggesting this because I was reading a function `handle_compare` and there was just a bit too much time required to remember what the function actually produces and how to use the output. This would allow us to reference the functions more easily and know what the inputs and outputs of a function are and more easily reuse them heavily.

Reply:
Yep, I've added comments. I'm still not really happy with these functions but I want to minimise the actual code changed in this PR, but will tackle those in the next ones